### DISCUSSION TITLE
outputting flux (volume flux and mass flux) [PorousFlow]

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#19235] [Q&A Modules: General] [answered]
Hi guys,

just wondering, is there a MOOSE object that can calculate flux within the domain?

I see we have `PorousFlowDarcyVelocityComponent` which outputting the velocity via AuxVariable/AuxKernel
https://mooseframework.inl.gov/source/auxkernels/PorousFlowDarcyVelocityComponent.html

but what about volume flux (m3/s) or mass flux (kg/s)? I'm aware that it might need to be calculated at the surfaces of each element since it has 'A' in its equation.

Do I need to write my own kernel/object for this? but it's quite standard, I believe MOOSE must have it somewhere.

Kind regards,
Traiwit

**Comment by GiudGiud** [comment]
I d use a VolumetricFlowRate from the Navier Stokes module for an external (to the variable) boundary
https://mooseframework.inl.gov/source/postprocessors/VolumetricFlowRate.html
For an internal sideset
https://mooseframework.inl.gov/source/postprocessors/InternalVolumetricFlowRate.html

EDIT
For porous flow, does any of the postprocessors listed here work for you:
https://mooseframework.inl.gov/modules/porous_flow/additional_objects.html

seems https://mooseframework.inl.gov/source/postprocessors/PorousFlowPlotQuantity.html can measure fluxes

Guillaume

**Reply by GiudGiud** [reply]
There s also a diffusive Flux posptrocessor. Depends on what fluxes you have. 
Since you are using porous flow, the porous flow postprocessors should be tried first

**Reply by Traiwit** [reply]
Hi @GiudGiud,

Thank you for pointing out all these helpful objects.
Unfortunately, they don't seem to be what I wanted, I think I need to discuss with my team further before proceeding on this.

The best bet right now is to take the Darcy vel at each element and multiply it by the average surface area of each element.

I will get back if I have a better idea about this.

Kind regards,
Traiwit

**Comment by WilkAndy** [comment, marked as answer]
It depends on the BC you're using on that boundary (if any: perhaps there are none because it's an internal boundary or because it's a no-flow boundary)

- if one of the [PorousFlowSinks](https://mooseframework.inl.gov/source/bcs/PorousFlowSink.html) then you use a `save_in` to save the contribution to an AuxVariable, then a `NodalSum` to sum all the contributions.  There are examples of this in the test suite.
- Otherwise, you can probably use a [PorousFlowOutflowBC](https://mooseframework.inl.gov/source/bcs/PorousFlowOutflowBC.html) and a `save_in` to record the flux.   I'm sorry that this is rather convoluted, and i've been meaning to address this for some time, but to record the flux you'll probably need *two* BCs of this type, one with `multiplier = 1` and one with `multiplier = -1`.  That means the contributions to the residual will sum to zero (these BCs will actually "do nothing") but you can `save_in` for the `multiplier = 1` version only, hence recording the flux.  Ugly.... sorry.

a

**Reply by Traiwit** [reply]
Thanks @WilkAndy, yes I'm using one of PorousFlowSinks, I will give it a go.
So I should expect the results to be the total flux of the particular surface, right?

Kind regards,
Traiwit

**Reply by WilkAndy** [reply]
Yes, that is correct.  The NodalSum will be the flux by definition.  Have a look at `porous_flow/test/tests/sinks/PorousFlowPiecewiseLinearSink_BC_eg1.i` for an example.

**Reply by Nuriacor** [reply]
@WilkAndy does `InternalVolumetricFlowRate` work in PorousFlow? If not, is there something similar?
